api/daemon/daemon_test.go

```diff
@@ -0,0 +1,79 @@
+package daemon
+
+import (
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/quintilesims/layer0/api/lock/mock_lock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestDaemonRun(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockLock := mock_lock.NewMockLock(ctrl)
+	mockLock.EXPECT().
+		Acquire("lock_id").
+		Return(true, nil)
+
+	mockLock.EXPECT().
+		Release("lock_id").
+		Return(nil)
+
+	var called bool
+	daemon := NewDaemon("", "lock_id", mockLock, func() error {
+		called = true
+		return nil
+	})
+
+	daemon.Run()
+	assert.True(t, called)
+}
+
+func TestDaemonHonorsLock(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockLock := mock_lock.NewMockLock(ctrl)
+	mockLock.EXPECT().
+		Acquire("lock_id").
+		Return(false, nil)
+
+	daemon := NewDaemon("", "lock_id", mockLock, func() error {
+		t.Fatalf("The Daemon's function was called")
+		return nil
+	})
+
+	daemon.Run()
+}
+
+func TestDaemonRunEvery(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockLock := mock_lock.NewMockLock(ctrl)
+	mockLock.EXPECT().
+		Acquire(gomock.Any()).
+		Return(true, nil).
+		AnyTimes()
+
+	mockLock.EXPECT().
+		Release(gomock.Any()).
+		Return(nil).
+		AnyTimes()
+
+	c := make(chan bool)
+	daemon := NewDaemon("", "", mockLock, func() error {
+		c <- true
+		return nil
+	})
+
+	ticker := daemon.RunEvery(time.Nanosecond)
+	defer ticker.Stop()
+
+	for i := 0; i < 5; i++ {
```

Comment:
I think it's wise to avoid `time.Sleep` during unit tests as much as possible. Also, wouldn't that essentially be testing the [ticker](https://golang.org/pkg/time/#Ticker) object?

Reply:
Not necessarily. From a caller's perspective of RunEvery, I have no idea how it is internally scheduling calls (whether it is a ticker or not). Just that RunEvery, will run every x duration. I would think the test should reflect that.

The test here is just testing RunEvery will run things a bunch of arbitrary times based on the smallest unit passed in for duration. The 5 can even be 2 and would still pass the test.

Reply:
> From a caller's perspective of RunEvery, I have no idea how it is internally scheduling calls (whether it is a ticker or not)

I don't know if I agree with that, because the function returns the ticker object it uses 
```
func (d *Daemon) RunEvery(period time.Duration) *time.Ticker
```

Reply:
I've updated this test to ensure the `daemonFN` doesn't run after `ticker.Stop()` is called; I believes reasonably tests that the ticker is indeed the scheduling mechanism .